### PR TITLE
Blaze: Connect Blaze Ad Destination with Blaze Campaign Form

### PR DESCRIPTION
**Pull request by hafizrahman**
Part of #11609 

## Description

This PR enables data passing between Blaze Ad destination screen and Blaze campaign form.

Unit tests are to be added separately in a new PR.

## Testing instructions

1. Start with a Blaze eligible site,
2. Go to a Product, then tap "Promote with Blaze",
3. On the campaign form screen, scroll down to "Ad destination" option and tap it,
4. On the Ad Destination screen, ensure that the "Product URL" is set as Destination URL as the default,
5. Make changes (change destination URL, add URL parameters), then tap Save.
6. Ensure you're returned to campaign form screen.
7. Tap "Ad destination" again.
8. On the Ad Destination screen, ensure that the changes from step 5 are preserved and shown.

### CODE REVIEW

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot is included in the pull request. Consider adding some for clarity.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

**Comment by selanthiraiyan** [comment]
@hafizrahman Could you kindly resolve the conflicts? Thank you!

**Review by selanthiraiyan — COMMENTED** [review]
@hafizrahman Could you kindly resolve the conflicts? Thank you!

**Comment by hafizrahman** [comment]
@selanthiraiyan 

> Could you kindly resolve the conflicts? Thank you!

Will do! (The conflict wasn't there when I published the PR yesterday 😅 )

File: WooCommerce/Classes/ViewRelated/Blaze/AdDestination/BlazeAdDestinationSettingView.swift [hunk 1]
```diff
@@ -10,8 +10,12 @@ struct BlazeAdDestinationSettingView: View {
 
     @State private var isShowingAddParameterView = false
 
-    init(viewModel: BlazeAdDestinationSettingViewModel) {
+    private let onDismiss: () -> Void
+
+    init(viewModel: BlazeAdDestinationSettingViewModel,
+         onDismiss: @escaping () -> Void ) {
```
Comment (selanthiraiyan) [inline WooCommerce/Classes/ViewRelated/Blaze/AdDestination/BlazeAdDestinationSettingView.swift]: I don't think we need the `onDismiss` closure. I can see that the view model already has `onSave` closure for passing back the value.

File: WooCommerce/Classes/ViewRelated/Blaze/CampaignCreation/BlazeCampaignCreationForm.swift [hunk 2]
```diff
@@ -232,7 +240,7 @@ private extension BlazeCampaignCreationForm {
         .padding(.vertical, Layout.contentPadding)
     }
 
-    func detailView(title: String, content: String, action: @escaping () -> Void) -> some View {
+    func detailView(title: String, content: String, singleLineSubtitle: Bool = false, action: @escaping () -> Void) -> some View {
```
Comment (selanthiraiyan) [inline WooCommerce/Classes/ViewRelated/Blaze/CampaignCreation/BlazeCampaignCreationForm.swift]: nit: Consider renaming this into `isContentSingleLine` to match the other `content` parameter. 
```suggestion
    func detailView(title: String, content: String, isContentSingleLine: Bool = false, action: @escaping () -> Void) -> some View {
```

File: WooCommerce/Classes/ViewRelated/Blaze/AdDestination/BlazeAdUrlParameter.swift [hunk 3]
```diff
@@ -21,3 +21,12 @@ extension Array where Element == BlazeAdURLParameter {
         return self.map { "\($0.key)=\($0.value)" }.joined(separator: "&")
     }
 }
+
+/// Convert a URLComponents's query items to BlazeAdURLParameter array.
+extension URLComponents {
+    func toBlazeAdURLParameters() -> [BlazeAdURLParameter] {
```
Comment (selanthiraiyan) [inline WooCommerce/Classes/ViewRelated/Blaze/AdDestination/BlazeAdUrlParameter.swift]: Adding this as a public method to an extension of `URLComponents` is unnecessary if we don't plan to use this helper in other places of the codebase. For comparison, if you consider [this other extension](https://github.com/woocommerce/woocommerce-ios/blob/trunk/WooCommerce/Classes/ViewRelated/Blaze/AdDestination/BlazeAdUrlParameter.swift#L19) it is only applicable to `BlazeAdURLParameter` which makes sense. The Xcode autocomplete will list the method only if the array contains `BlazeAdURLParameter` elements.

If you plan to use it only in `BlazeAdDestinationSettingViewModel`, you could move this extension to `BlazeAdDestinationSettingViewModel` and mark it as `private`.

File: WooCommerce/Classes/ViewRelated/Blaze/CampaignCreation/BlazeCampaignCreationFormViewModel.swift [hunk 4]
```diff
@@ -105,11 +105,30 @@ final class BlazeCampaignCreationFormViewModel: ObservableObject {
...
+        if let productURL = productURL, let siteURL = siteURL {
+            BlazeAdDestinationSettingViewModel(
+                productURL: productURL,
+                homeURL: siteURL,
+                finalDestinationURL: finalDestinationURL) { [weak self] finalDestinationURL in
+                    self?.finalDestinationURL = finalDestinationURL
+                }
+        } else {
+            nil
+        }
```
Comment (selanthiraiyan) [inline WooCommerce/Classes/ViewRelated/Blaze/CampaignCreation/BlazeCampaignCreationFormViewModel.swift]: Consider using `guard` to exit early to improve readability. Also, log an error when these URLs are empty.

File: WooCommerce/Classes/ViewRelated/Blaze/AdDestination/BlazeAdDestinationSettingView.swift [hunk 5]
```diff
@@ -79,7 +83,14 @@ struct BlazeAdDestinationSettingView: View {
                     Button(Localization.cancel) {
                         dismiss()
                     }
-                    .foregroundColor(Color(uiColor: .accent))
+                }
+
+                ToolbarItem(placement: .confirmationAction) {
+                    Button(Localization.save) {
+                        viewModel.confirmSave()
+                        onDismiss()
```
Comment (selanthiraiyan) [inline WooCommerce/Classes/ViewRelated/Blaze/AdDestination/BlazeAdDestinationSettingView.swift]: You can use the `dismiss` method to dismiss the view. I don't think you need the `onDismiss` closure is necessary. 
```suggestion
                        dismiss()
```

File: WooCommerce/Classes/ViewRelated/Blaze/AdDestination/BlazeAdDestinationSettingViewModel.swift [hunk 6]
```diff
@@ -66,20 +71,48 @@ final class BlazeAdDestinationSettingViewModel: ObservableObject {
...
-        self.selectedDestinationType = selectedDestinationType
-        self.parameters = parameters
+        self.initialFinalDestinationURL = finalDestinationURL
+        self.onSave = onSave
+
+        setStartingDestinationType()
+        setStartingParameters()
+    }
+
+    private func setStartingDestinationType() {
```
Comment (selanthiraiyan) [inline WooCommerce/Classes/ViewRelated/Blaze/AdDestination/BlazeAdDestinationSettingViewModel.swift]: You can move this and the other private methods to a separate `private` extension to make is easier to read through the available public methods.

File: WooCommerce/Classes/ViewRelated/Blaze/AdDestination/BlazeAdDestinationSettingViewModel.swift [hunk 7]
```diff
@@ -66,20 +71,48 @@ final class BlazeAdDestinationSettingViewModel: ObservableObject {
...
+        if initialFinalDestinationURL.hasPrefix(productURL) {
+            selectedDestinationType = .product
+        } else {
+            selectedDestinationType = .home
+        }
+    }
+
+    private func setStartingParameters() {
```
Comment (selanthiraiyan) [inline WooCommerce/Classes/ViewRelated/Blaze/AdDestination/BlazeAdDestinationSettingViewModel.swift]: Opinion: What do you think about renaming these methods into `initializeParameters` or `initParameters`? Same for the other `setStartingDestinationType` method. IMO, `starting` might be confusing.

**Review by selanthiraiyan — COMMENTED** [review]
Thanks for the work, Hafiz! This works great. 🎉 

I left a few nits, comments and suggestions about the code. Please take a look and let me know your thoughts.

File: WooCommerce/Classes/ViewRelated/Blaze/AdDestination/BlazeAdDestinationSettingViewModel.swift (hunk 6 above)
Comment (hafizrahman) [reply WooCommerce/Classes/ViewRelated/Blaze/AdDestination/BlazeAdDestinationSettingViewModel.swift]: Interesting idea for grouping private methods that way, so far I assumed extensions are more for grouping methods based on similar concepts/tasks. I agree this makes the class easier to read between private vs public methods.

File: WooCommerce/Classes/ViewRelated/Blaze/AdDestination/BlazeAdDestinationSettingViewModel.swift (hunk 7 above)
Comment (hafizrahman) [reply WooCommerce/Classes/ViewRelated/Blaze/AdDestination/BlazeAdDestinationSettingViewModel.swift]: I agree! Renamed as part of https://github.com/woocommerce/woocommerce-ios/pull/11703/commits/206937a67ac42e68eb2a3ec052fe35b84ba08030

File: WooCommerce/Classes/ViewRelated/Blaze/AdDestination/BlazeAdDestinationSettingViewModel.swift (hunk 6 above)
Comment (hafizrahman) [reply WooCommerce/Classes/ViewRelated/Blaze/AdDestination/BlazeAdDestinationSettingViewModel.swift]: Handled in https://github.com/woocommerce/woocommerce-ios/pull/11703/commits/206937a67ac42e68eb2a3ec052fe35b84ba08030

File: WooCommerce/Classes/ViewRelated/Blaze/CampaignCreation/BlazeCampaignCreationFormViewModel.swift (hunk 4 above)
Comment (hafizrahman) [reply WooCommerce/Classes/ViewRelated/Blaze/CampaignCreation/BlazeCampaignCreationFormViewModel.swift]: Great suggestions. Done in https://github.com/woocommerce/woocommerce-ios/pull/11703/commits/557244948176af6ede6fd5c8028add568a7fc393

File: WooCommerce/Classes/ViewRelated/Blaze/AdDestination/BlazeAdDestinationSettingView.swift (hunk 1 above)
Comment (hafizrahman) [reply WooCommerce/Classes/ViewRelated/Blaze/AdDestination/BlazeAdDestinationSettingView.swift]: Handled in https://github.com/woocommerce/woocommerce-ios/pull/11703/commits/6968fb8167919165f616fecb79b3259d5a46c445

File: WooCommerce/Classes/ViewRelated/Blaze/AdDestination/BlazeAdUrlParameter.swift (hunk 3 above)
Comment (hafizrahman) [reply WooCommerce/Classes/ViewRelated/Blaze/AdDestination/BlazeAdUrlParameter.swift]: Good point. At first I wasn't sure if this is going to be used in multiple places or not. At first I had the thought that the campaign form might need this too. Later I decided for the conversion to only be handled in the ad destination setting, but then did not revisit this.

Comment (hafizrahman) [reply WooCommerce/Classes/ViewRelated/Blaze/AdDestination/BlazeAdUrlParameter.swift]: Done https://github.com/woocommerce/woocommerce-ios/pull/11703/commits/b5c639d04a81ed9f932eee71830560e58d21cb83

File: WooCommerce/Classes/ViewRelated/Blaze/AdDestination/BlazeAdDestinationSettingViewModel.swift [hunk 8]
```diff
@@ -148,3 +182,12 @@ private extension BlazeAdDestinationSettingViewModel {
         )
     }
 }
+
+/// Convert a URLComponents's query items to BlazeAdURLParameter array.
+private extension URLComponents {
+    func toBlazeAdURLParameters() -> [BlazeAdURLParameter] {
+        guard let queryItems = self.queryItems else { return [] }
```
Comment (selanthiraiyan) [inline WooCommerce/Classes/ViewRelated/Blaze/AdDestination/BlazeAdDestinationSettingViewModel.swift]: 💡 The guard statement can be shortened into
```suggestion
        guard let queryItems else { return [] }
```

**Review by selanthiraiyan — APPROVED** [review]
Thanks for the updates, Hafiz! 🚀

**Pull request merged**